gpii/node_modules/preferencesServer/src/preferencesServer.js

```diff
@@ -36,116 +37,188 @@ fluid.defaults("gpii.preferencesServer", {
+            "target": "{that authorizationService}.options.components.dataStore"
+        },
+        preferencesServiceDataStore: {
+            "record": "{gpii.preferencesServer}.dataStore",
+            "target": "{that preferencesService}.options.components.dataStore"
+        }
     }
 });
 
-gpii.preferencesServer.getRawPreferences = function (rawPreferencesDataSource, userToken) {
+/**
+ * Simple function to get the 'view'. In case none is given, defaults to the "flat" ontology.
+ * @param {String} view - A view.
+ * @return {String} The value of the 'view', or "flat" if none is given.
+ */
+gpii.preferencesServer.getView = function (view) {
+    return view ? view : "flat";
+};
+
+/**
+ * The API to get preferences. Note that this API returns the entire value of the prefsSafe.preferences
+ * rather than the particular prefs set that is associated with the GPII key provided as an argument.
+ * @param {Component} preferencesService - An instance of gpii.preferencesServer.preferencesService component.
+ * @param {Component} ontologyHandler - An instance of gpii.ontologyHandler component.
+ * @param {String} gpiiKey - An GPII key.
+ * @param {String} view - An view that the preferences should be transformed upon.
+ * @return {Promise} A promise object whose resolved value is the preferences transformed based on the view.
+ */
+gpii.preferencesServer.getPreferences = function (preferencesService, ontologyHandler, gpiiKey, view) {
+    fluid.log("gpii.preferencesServer.getPreferences called - fetching preferences for the GPII key " + gpiiKey);
+    var rawPrefsPromise = preferencesService.getPreferencesByGpiiKey(gpiiKey);
+    var mapper = function (rawPrefs) {
+        var prefs = rawPrefs ? ontologyHandler.rawPrefsToOntology(rawPrefs, view) : undefined;
+        fluid.log("Preferences Server, getPreferences(), returning preferences: ", prefs);
+        return prefs;
+    };
+    return fluid.promise.map(rawPrefsPromise, mapper);
+};
+
+/**
+ * The API to create preferences. This API does:
+ * 1. If the GPII key is not provided, first auto generat a GPII key and create it, then create a new prefs safe with the given preferences and associate this prefs safe with the new GPII key;
+ * 2. If the GPII key is provided but not yet exists, create it first, then create a new prefs safe with the given preferences and associate the prefs safe with the GPII key;
+ * 3. If the GPII key is provided and already exists, reject with an error;
+ * 4. If preferences is not provided, reject with an error.
+ * @param {Component} preferencesService - An instance of gpii.preferencesServer.preferencesService component.
+ * @param {Component} ontologyHandler - An instance of gpii.ontologyHandler component.
+ * @param {Object} preferences - A preferences object.
+ * @param {String} view - An view that the provided preferences is based upon.
+ * @param {String} gpiiKey - (Optional) The GPII key to be created.
```

Comment:
Optional params are signalled in JSDocs by putting the parameter name in square brackets - http://usejsdoc.org/tags-param.html#optional-parameters-and-default-values - although we tend to also retain the text [optional] for clarity in the description